### PR TITLE
UI/Qt: Don't show URL when a new tab is initially focused

### LINKED ISSUE #137: Select URL when opening new tab?
When I open a new tab, I do so because I want to type a URL or search. Currently, this is somewhat annoying in Ladybird because the location field already contains "about:newtab".

Firefox seems to deal with this by leaving the location field blank when opening new tabs, but automatically selecting its contents would also avoid the issue.

This patch seems to work:
```patch
diff --git a/Ladybird/Qt/Tab.cpp b/Ladybird/Qt/Tab.cpp
index d00db16a71..955120c4af 100644
--- a/Ladybird/Qt/Tab.cpp
+++ b/Ladybird/Qt/Tab.cpp
@@ -156,6 +156,9 @@ Tab::Tab(BrowserWindow* window, WebContentOptions const& web_content_options, St
 
     view().on_url_change = [this](auto const& url) {
         m_location_edit->setText(qstring_from_ak_string(url.serialize()));
+
+        if (m_location_edit->hasFocus())
+            m_location_edit->selectAll();
     };
 
     QObject::connect(m_location_edit, &QLineEdit::returnPressed, this, &Tab::location_edit_return_pressed);
```

### PR DESCRIPTION
**Pull request by tcl3**
Previously, I often found myself in this situation:

* Copy a URL from somewhere
* Open a new tab Ladybird
* Paste the new URL and hit Enter
* `about:newtabhttp://some-site-i-want-to-visit.com not found` :sob: 

This PR makes 2 separate but related changes that make it easier for the user to copy and paste URLs into a new tab.  

Firstly, the `about:newtab` URL is no longer displayed in the location bar when viewing the new tab page. This also has the effect of allowing the user to see the placeholder text when opening a new tab

~~Secondly, the "Page on New Tab" option has been replaced with a "Page on New Window" option. The `about:newtab` page is now shown when opening a new tab in an existing window. This matches the behavior of all other browsers that I have used - Firefox also allows the user to choose between `about:blank` and `about:newtab` for new tabs, I'd be happy to add that behavior if desired.~~
**EDIT:** I've removed this change for now, see the discussion below.

I haven't updated the AppKit chrome with this new behavior as part of this PR.

Demo with "Page on New Window" set to "http://example.com":

https://github.com/LadybirdBrowser/ladybird/assets/2817754/37bde33b-73a9-42be-b7e6-2a5c1bba3749

Fixes #137

### CODE REVIEW

**Comment by awesomekling** [comment]
Hmm, my Chromium install definitely opens something other than about:blank in every new tab. So it doesn't seem accurate to say that no other browsers have a "new tab" page.

Or maybe I'm misunderstanding something?

**Comment by tcl3** [comment]
> Hmm, my Chromium install definitely opens something other than about:blank in every new tab. So it doesn't seem accurate to say that no other browsers have a "new tab" page.
> 
> Or maybe I'm misunderstanding something?

AFAICT, the default behavior in Chrome is to display `about:newtab` in every new tab, which is the same behavior I've implemented here. 

This could well be something I've done to the Chrome install on my machine though.

**Comment by awesomekling** [comment]
> > Hmm, my Chromium install definitely opens something other than about:blank in every new tab. So it doesn't seem accurate to say that no other browsers have a "new tab" page.
> > Or maybe I'm misunderstanding something?
> 
> AFAICT, the default behavior in Chrome is to display `about:newtab` in every new tab, which is the same behavior I've implemented here.
> 
> This could well be something I've done to the Chrome install on my machine though.

But.. I (and probably others) definitely never want to see the "new tab" page at all. And you're now forcing us to see it :sweat_smile:

**Review by trflynn89 — COMMENTED** [review]
I'm also not sure that "page on new window" makes sense as an option. I would just expect a "page on new tab" option to apply to both new tabs and windows

**Comment by AtkinsSJ** [comment]
> I'm also not sure that "page on new window" makes sense as an option. I would just expect a "page on new tab" option to apply to both new tabs and windows

I think Firefox might have the distinction of showing a "home page" if you open it fresh, but a different page if you open a new tab in an existing process. Though, it's been a long time since I've opened it fresh so I might not be remembering. That's different than new tab/new window though.

**Comment by Zaggy1024** [comment]
To solve the copy/paste problem, why not instead just select all the text in the location bar upon opening a new tab or window? Then, typing or pasting will replace all the text, but one can still see the URL being used for the new tab.

**Comment by circl-lastname** [comment]
> To solve the copy/paste problem, why not instead just select all the text in the location bar upon opening a new tab or window? Then, typing or pasting will replace all the text, but one can still see the URL being used for the new tab.

That seems to already be the behavior

**Comment by jamierocks** [comment]
> No it's the same I'm pretty sure

This is wrong - they are separate.

image

**Comment by circl-lastname** [comment]
Ohh, I just mean Firefox doesn't have a specifically different default home vs newtab page

**Comment by hughdavenport** [comment]
I think this is adding too many changes which is causing confusion.

I would expect a change like this to just hide the about:newtab page, which **is** similar to other open source browsers. Currently other browsers do the following:

Chromium
 - new tab and new window both do the same, which is open chromium's `about:newtab`/`chrome://newtab`.
 - when opened, the address bar is focused, and is blank (not even the placeholder)
 - clicking off the address bar will show the placeholder text.
 - this page shows the current search engine configured, and frequently used pages (can be hidden)
 - it can configure the page(s) to use when *starting up*, but not for `about:newtab` seemingly.

Firefox
- new tab and new window can both be configured independently
- new window (also browser startup page) can be configured to either `Firefox Home`, a custom page, or `about:blank`
- new tab can only be configured to either `Firefox Home` or `about:blank` (changing this changes `about:newtab`)
- `Firefox Home` can be configured to show any of `Search` (with configured search engine), `shortcuts`, and `recent activity`
- when opened, both `Firefox Home` and `about:blank` have the address bar focused, and is showing the placeholder text (essentially cleared ready to type a URL).

Ladybird:
- Can configure a page for new tabs (and windows).
- if using a custom address, when a new tab is opened the address bar is not focused
- if using `about:blank` or `about:newtab`, when a new tab is opened the bar is focused with the cursor at the end of `about:....`
- if using `about:blank` or `about:newtab`, when a new window is opened (or ladybird started) the address bar is **not** focused

In summary, I think the other open source browsers currently give a better user experience and having the following **single** change to ladybird would improve that
- if page is `about:blank` or `about:newtab` the location bar should be clear and be focused when first opened.

Currently this PR adds a bunch more that isn't needed to fulfil that one basic change to improve user experience, and should probably get split into multiple PRs. Happy for any of the maintainers to correct me on this :)

**Comment by hughdavenport** [comment]
I can provide some screenshots and screencasts to show what I have experienced if my above summary did not clarify fully :)

File: Ladybird/Qt/LocationEdit.cpp [hunk 1]
```diff
@@ -109,4 +109,24 @@ void LocationEdit::highlight_location()
     QCoreApplication::sendEvent(this, &event);
 }
 
+static bool should_hide_url(StringView serialized_url)
```
[the remark this answers is not in the document]
Comment (tcl3) [reply Ladybird/Qt/LocationEdit.cpp]: Good point! I've moved this into LibWebView.

**Comment by tcl3** [comment]
Thanks for the feedback on this guys! What I've done for now is drop the second change from this PR, as it seems a consensus still needs to be reached on the best approach to take here. 

> I think this is adding too many changes which is causing confusion.

That's a fair point - he reason that I decided to make both these changes together was to ensure a consistent experience regardless of the user's settings. With just the `about:newtab` change, the URL copy/paste problem that I mentioned initially still occurs when the user has a non-default new tab page. I think hiding `about:newtab` is definitely an improvement on the old behavior though and the change is worth making on its own.

> To solve the copy/paste problem, why not instead just select all the text in the location bar upon opening a new tab or window? Then, typing or pasting will replace all the text, but one can still see the URL being used for the new tab.

My first thought when looking at this problem was to do this. The issue I ran into was that the focus is lost when the location bar text is changed, which happens quite often when a site is loaded because of URL redirections. Obviously this problem is fixable - an easy thing to do would be to not change location bar text while the location bar is focused, for example. I thought it easier to sidestep the problem entirely though, by doing something similar to what other browsers do. 

I've also added `about:blank` to the list of URLs to not show in the location bar. It's worth noting that not all browsers do this - Chrome for example, doesn't hide `about:blank`. The change seems pretty harmless to me though.

**Comment by tcl3** [comment]
So I think I've come up with a solution that addresses everyone's concerns. The URL is now not shown when the location bar is initially focused on a newly activated tab. The URL is shown when the location bar loses focus and the user hasn't entered anything into it.

This solution fixes the URL copy pasting problem, allows the user to see the placeholder text after opening a new tab and doesn't obscure the current URL apart from when the user is entering a new address.

Video:

https://github.com/LadybirdBrowser/ladybird/assets/2817754/1b30f9ad-eb80-4bee-b51e-e5b8be369920

**Comment by awesomekling** [comment]
All right, let's roll with this for now :)
